gtest-core-metacling-test-TClingLoadUnloadFile fails sporadically

We disabled that test in https://github.com/root-project/root/pull/14281

Relevant test failure output:
```
   ROOT::TestSupport::ForbidDiagnostics::handler(): Diagnostic in 'TUnixSystem::ACLiC':
  creating shared library /github/home/ROOT-CI/build/core/metacling/test/concurrent_load_unload_one_lib_per_thread_4_cpp.so
  terminate called after throwing an instance of 'std::bad_alloc'
    what():  std::bad_alloc
  CMake Error at /github/home/ROOT-CI/src/cmake/modules/RootTestDriver.cmake:232 (message):
    error code: Subprocess aborted
```

--- a/core/metacling/test/TClingLoadUnloadFileTests.cxx
+++ b/core/metacling/test/TClingLoadUnloadFileTests.cxx
@@ -75,9 +75,6 @@ void remove_n_library_artifacts(const std::string &basename, unsigned int n)
 
 TEST(TClingLoadUnloadFile, ConcurrentLoadUnloadSameLib)
 {
-   GTEST_SKIP() << "fails sporadically, see "
-                << "https://github.com/root-project/root/issues/14121";
-
    ROOT::EnableThreadSafety();
 
    std::string basename{"concurrent_load_unload_same_lib"};
@@ -102,9 +99,6 @@ TEST(TClingLoadUnloadFile, ConcurrentLoadUnloadSameLib)
 
 TEST(TClingLoadUnloadFile, ConcurrentLoadUnloadOneLibPerThread)
 {
-   GTEST_SKIP() << "fails sporadically, see "
-                << "https://github.com/root-project/root/issues/14121";
-
    ROOT::EnableThreadSafety();
 
    std::string basename{"concurrent_load_unload_one_lib_per_thread"};

--- a/interpreter/cling/include/cling/MetaProcessor/MetaSema.h
+++ b/interpreter/cling/include/cling/MetaProcessor/MetaSema.h
@@ -37,7 +37,6 @@ namespace cling {
     bool m_IsQuitRequested;
 
     llvm::DenseMap<const clang::FileEntry*, const Transaction*> m_FEToTransaction;
-    llvm::DenseMap<const Transaction*, const clang::FileEntry*> m_TransactionToFE;
   public:
     enum SwitchMode {
       kOff = 0,

--- a/interpreter/cling/lib/MetaProcessor/MetaSema.cpp
+++ b/interpreter/cling/lib/MetaProcessor/MetaSema.cpp
@@ -213,14 +213,23 @@ namespace cling {
       return AR_Success;
 
     const Transaction* unloadPoint = (*TI).second;
+    // Erase the entry before reverting: as of now the file is no longer
+    // registered as loaded, whatever the outcome of the reversion.
+    m_FEToTransaction.erase(TI);
     if (interpreterHasTransaction(m_Interpreter, unloadPoint)) {
       // Revert all the transactions that came after `unloadPoint'.
       while (m_Interpreter.getLastTransaction() != unloadPoint) {
-        if (const auto ThisFE = m_TransactionToFE[m_Interpreter.getLastTransaction()]) {
-          auto I = m_FEToTransaction.find(ThisFE);
-          if (I != m_FEToTransaction.end())
-            m_FEToTransaction.erase(I);
-        }
+        const Transaction* T = m_Interpreter.getLastTransaction();
+        // Any file whose unload point is `T' can no longer be reverted: drop
+        // its registration. This must not be left behind: `T' is deallocated
+        // during the reversion and its address may be reused by a future
+        // transaction, which would alias the stale entry.
+        llvm::SmallVector<const clang::FileEntry*, 2> revertedFiles;
+        for (const auto &Entry : m_FEToTransaction)
+          if (Entry.second == T)
+            revertedFiles.push_back(Entry.first);
+        for (const clang::FileEntry* RFE : revertedFiles)
+          m_FEToTransaction.erase(RFE);
         m_Interpreter.unload(/*numberOfTransactions=*/1);
       }
 
@@ -231,7 +240,6 @@ namespace cling {
       m_MetaProcessor.getOuts() << "!!!ERROR: Transaction for file: " << file
                                 << " has already been unloaded\n";
     }
-    m_FEToTransaction.erase(TI);
     return AR_Success;
   }
 
@@ -495,9 +503,7 @@ namespace cling {
       return;
     }
 
-    if (*FE && !m_FEToTransaction[*FE]) {
+    if (*FE && !m_FEToTransaction[*FE])
       m_FEToTransaction[*FE] = unloadPoint;
-      m_TransactionToFE[unloadPoint] = *FE;
-    }
   }
 } // end namespace cling

